cmd/cmdoptions/metrics.go

```diff
@@ -16,49 +17,113 @@ import (
 )
 
 type metricsHandler struct {
-	registry *prometheus.Registry
-	tags     map[string]string
+	metrics *Metrics
+	labels  []string
+	values  []string
 }
 
 var _ client.MetricsHandler = (*metricsHandler)(nil)
 
 func (h *metricsHandler) WithTags(tags map[string]string) client.MetricsHandler {
 	// Make enough space for the handlers tags which are populated first
-	mergedTags := make(map[string]string, len(h.tags))
-	for t, v := range h.tags {
-		mergedTags[t] = v
+	mergedTags := make(map[string]string, len(h.labels))
+	for i, l := range h.labels {
+		mergedTags[l] = h.values[i]
+	}
+	for l, v := range tags {
+		mergedTags[l] = v
+	}
+
+	var labels, values []string
+	for l, v := range mergedTags {
+		labels = append(labels, l)
+		values = append(values, v)
 	}
-	for t, v := range tags {
-		mergedTags[t] = v
+
+	return &metricsHandler{
+		metrics: h.metrics,
+		labels:  labels,
+		values:  values,
 	}
-	return &metricsHandler{registry: h.registry, tags: mergedTags}
 }
 
-func (h *metricsHandler) mustRegisterIgnoreDuplicate(c prometheus.Collector) {
-	err := h.registry.Register(c)
-	var alreadyRegisteredError prometheus.AlreadyRegisteredError
-	if err != nil && !errors.As(err, &alreadyRegisteredError) {
-		panic(err)
+func (h *metricsHandler) getOrCreateCounter(name string) *prometheus.CounterVec {
```

Comment:
Seems like you might as well inline these get or create functions since they're 1:1 used by the public functions